ethash/src/cache.rs

```diff
@@ -0,0 +1,282 @@
+	pub fn cache_path(&self) -> PathBuf {
+		cache_path(self.cache_dir.as_ref(), &self.ident)
+	}
+
+	pub fn flush(&mut self) -> io::Result<()> {
+		let path = self.cache_path();
+
+		if let Some(last) = self.epoch.checked_sub(2).map(|ep| {
+			cache_path(self.cache_dir.as_ref(), &self.builder.epoch_to_ident(ep))
+		})
+		{
+			fs::remove_file(last)?;
+		}
+
+		consume_cache(&mut self.cache, &path)
+	}
+}
+
+fn make_memmapped_cache<P: AsRef<Path>>(
+	path: &P,
+	num_nodes: usize,
+	ident: &H256,
+) -> io::Result<Mmap> {
+	use std::fs::OpenOptions;
+
+	debug_assert_eq!(ident.len(), 32);
+
+	let file = OpenOptions::new().read(true).write(true).create(true).open(&path)?;
+	file.set_len((num_nodes * NODE_BYTES) as _)?;
+
+	let mut memmap = Mmap::open(&file, Protection::ReadWrite)?;
+
+	unsafe {
+		initialize_memory(memmap.mut_ptr() as *mut Node, num_nodes, ident);
+	}
+
+	Ok(memmap)
+}
+
+fn make_memory_cache(num_nodes: usize, ident: &H256) -> Vec<Node> {
+	let mut nodes: Vec<Node> = Vec::with_capacity(num_nodes);
+	// Use uninit instead of unnecessarily writing `size_of::<Node>() * num_nodes` 0s
+	unsafe {
+		nodes.set_len(num_nodes);
+		initialize_memory(nodes.as_mut_ptr(), num_nodes, ident);
+	}
+
+	nodes
+}
+
+fn cache_path<P: ToOwned<Owned = PathBuf> + ?Sized>(path: &P, ident: &H256) -> PathBuf
+where
+	PathBuf: Borrow<P>,
+{
+	let mut buf = path.to_owned();
+	buf.push(to_hex(ident));
+	buf
+}
+
+fn consume_cache<P: AsRef<Path>>(cache: &mut Cache, path: &P) -> io::Result<()> {
+	use std::fs::OpenOptions;
+
+	let new_cache = match *cache {
+		Either::Left(ref mut vec) => {
+			let mut file = OpenOptions::new().read(true).write(true).create(true).open(&path)?;
+
+			let buf = unsafe {
+				slice::from_raw_parts_mut(vec.as_mut_ptr() as *mut u8, vec.len() * NODE_BYTES)
+			};
+
+			println!("Writing...");
+			file.write(buf).map(|_| ())?;
+
+			file
+		}
+		Either::Right(ref mmap) => {
+			println!("Flushing...");
+			mmap.flush()?;
+
+			return Ok(());
+		}
+	};
+
+	// If creating the memmap fails, we keep the `Vec` and try again next time `flush` is called.
+	// This means that it's possible to use this on a system that doesn't support memory mapping at
+	// all, at the cost of higher RAM.
+	let memmap = Mmap::open(&new_cache, Protection::ReadWrite)?;
+	*cache = Either::Right(memmap);
+
+	Ok(())
+}
+
+fn cache_from_path<P: AsRef<Path>>(path: &P) -> io::Result<Cache> {
+	Mmap::open_path(path, Protection::ReadWrite)
+		.map(Either::Right)
+		.or_else(|_| read_from_path(path).map(Either::Left))
+}
+
+fn read_from_path<P: AsRef<Path>>(path: &P) -> io::Result<Vec<Node>> {
+	use std::fs::File;
+	use std::mem;
+
+	let mut file = File::open(path)?;
+
+	let mut nodes: Vec<u8> =
+		Vec::with_capacity(file.metadata().map(|m| m.len() as _).unwrap_or(1_000_000));
+	file.read_to_end(&mut nodes)?;
+	let out: Vec<Node> = unsafe {
+		Vec::from_raw_parts(
+			nodes.as_mut_ptr() as *mut _,
+			nodes.len() / NODE_BYTES,
+			nodes.capacity() / NODE_BYTES,
```

Comment:
fyi asserts should never fire, particularly not if the data coming from the user or filesystem happens to be "wrong". in this case, use error returns and bomb out cleanly if it's not trivially fixable/recoverable.

asserts are only there to encode logical tautologies and should always be accompanied with an explanation as to why they will *never* fire.